learn/building_programs/build_tools.md

```diff
@@ -4,75 +4,437 @@ title: Build tools
+action has to be performed.
+
+In the end, we will have a look at a complete ``Makefile``.
+
+```make
+# Disable all of make's implicit rules (similar to Fortran's implicit none)
+MAKEFLAGS += --no-builtin-rules --no-builtin-variables
+# configuration
+FC := gfortran
+LD := $(FC)
+RM := rm -f
+# list of all source files
+SRCS := tabulate.f90 functions.f90
+PROG := my_prog
+
+OBJS := $(addsuffix .o, $(SRCS))
+
+.PHONY: all clean
+all: $(PROG)
+
+$(PROG): $(OBJS)
+	$(LD) -o $@ $^
+
+$(OBJS): %.o: %
+	$(FC) -c -o $@ $<
+
+# define dependencies between object files
+tabulate.f90.o: functions.f90.o user_functions.mod
+
+# rebuild all object files in case this Makefile changes
+$(OBJS): $(MAKEFILE_LIST)
+
+clean:
+	$(RM) $(filter %.o, $(OBJS)) $(wildcard *.mod) $(PROG)
+```
+
+Since you are starting with ``make`` we highly recommend to always include
+the first line, like with Fortrans ``implicit none`` we do not want to have
+implicit rules messing up our ``Makefile`` in surprising and harmful ways.
+
+Next, we have a configuration section where we define variables, in case you
+want to switch out your compiler, it can be easily done here.
+We also introduced the ``SRCS`` variable to hold all source files, which is
+more intuitive than specifying object files, we can easily create the object
+files by appending a ``.o`` suffix using the functions ``addsuffix``.
+The ``.PHONY`` is a special rule, which should be used for all entry points
+of your ``Makefile``, here we define two entry point, we already know *all*,
+the new *clean* rule deletes all the build artifacts again such that we indeed
+start with a clean directory.
+
+Also, we slightly changed the build rule for the object files to account for
+appending the ``.o`` suffix instead of substituting it.
+Notice that we still need to explicitly define the interdependencies in the
+``Makefile``. We also added a dependency for the object files on the ``Makefile``
+itself, in case you change the compiler, this will allow you to savely rebuild.
+
+Now you know enough about ``make`` to use it for building small projects.
+
+> **Important**
+>
+> You might have noticed that ``make`` is not particularly easy to use and
+> it can be from time to time difficult to understand what is going
+> on under the hood.
+> In this guide, we avoided and disabled a lot of the commonly used ``make``
+> features that can be particularly troublesome if not used correctly, we highly
+> recommend staying away from them if you do not feel confident working with
+> ``make``.
+>
+> While ``make`` is indeed a handy tool to automate short interdependent
+> workflows and to build small projects, it should *never* be used to build
+> larger projects, like scientific programs.
+> In particular modern Fortran programs can hardly be handled by a ``make``
+> build system.
+>
+> To use ``make`` in a larger project, you cannot rely on ``make`` alone, usually
+> other tools are used to generate the ``Makefile`` completely or in parts.

```

Comment:
This section is strongly opinionated and without justification. It appears to reflect the author's experience with `make`; this may or may not be the case, but this is how it comes across in the text.

Any tool with advanced features will require further learning - warning off potential users implies that the tool is not fit for purpose, this is not something which we should imply for `make`.

I disagree with the absolute and generalised statement about the suitability of `make` for large projects. A balanced statement indicating the limitations of `make` and reasoning for using build configuration systems would suffice here IMO.

Reply:
I'm probably opinionated on make, I used it for quite a while extensively and now completely dropped it from all my projects because I couldn't make it scale well. I'll remove the paragraph than.

Reply:
Thanks for rewording - looks good now. I will try to do a more extensive review soon.